Fix some concurrency issues in ActivateBestChain()

skeees opened the pull request:
Originally this PR was just to add tests around concurrency in block validation - those tests seem to have uncovered another bug in ActivateBestChain - this now fixes that bug and adds tests.

ActivateBestChain (invoked after a new block is validated) proceeds in steps - acquiring and releasing cs_main while incrementally disconnecting and connecting blocks to sync to the most work chain known (FindMostWorkChain()). Every time cs_main is released the result of FindMostWorkChain() can change - but currently that value is cached across acquisitions of cs_main and only refreshed when an invalid chain is explored. It needs to be refreshed every time cs_main is reacquired. The test added in https://github.com/bitcoin/bitcoin/pull/13023/commits/6094ce73045fe0b4654ff94327c2059512af88fb will occasionally fail without the commit fixing this issue https://github.com/bitcoin/bitcoin/pull/13023/commits/26bfdbaddbb9f13864deb7241c6d513f22c5ab62


Original description below
--

After a bug discovered where UpdatedBlockTip() notifications could be triggered out of order (#12978), these unit tests check certain invariants about these signals.

The scheduler test asserts that a SingleThreadedSchedulerClient processes callbacks fully and sequentially.

The block validation test generates a random chain and calls ProcessNewBlock from multiple threads at random and in parallel. ValidationInterface callbacks verify that the ordering of BlockConnected BlockDisconnected and UpdatedBlockTip events occur as expected.

TheBlueMatt committed:
Do not unlock cs_main in ABC unless we've actually made progress.

Technically, some internal datastructures may be in an inconsistent
state if we do this, though there are no known bugs there. Still,
for future safety, its much better to only unlock cs_main if we've
made progress (not just tried a reorg which may make progress).

skeees committed:
Add unit tests for signals generated by ProcessNewBlock()

After a recent bug discovered in callback ordering in MainSignals,
this test checks invariants in ordering of
BlockConnected / BlockDisconnected / UpdatedChainTip signals

jamesob commented on src/Makefile.test.include:
```diff
@@ -87,7 +87,8 @@ BITCOIN_TESTS =\
   test/uint256_tests.cpp \
-  test/util_tests.cpp
+  test/util_tests.cpp \
+  test/validation_block_tests.cpp
```
Looks like there's an alphabetic ordering here that may be good to preserve.

jamesob commented on src/test/scheduler_tests.cpp:
```diff
@@ -112,4 +112,46 @@ BOOST_AUTO_TEST_CASE(manythreads)
+    }
+
+    // these are not atomic, if SinglethreadedSchedulerClient prevents
+    // parallel execution at the queue level no synchronization should required here
```
> should *be* required here

jamesob commented on src/test/validation_block_tests.cpp:
```diff
@@ -0,0 +1,158 @@
+    pblock->hashMerkleRoot = BlockMerkleRoot(*pblock);
+
+    while (!CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus()))
+        ++(pblock->nNonce);
```
Could be worth parameterizing `Block` with a `bool make_invalid` option to avoid the duplication here, but that's your call.

jamesob replied:
Yep, definitely agree.

jamesob commented on src/test/validation_block_tests.cpp:
```diff
@@ -0,0 +1,158 @@
+    return pblock;
+}
+
+void BuildChain(const uint256& root, int height, const unsigned int invalid_rate, const unsigned int branch_rate, const unsigned int max_size, std::vector<std::shared_ptr<const CBlock>>& blocks)
```
I know we don't have columnar limits in the styleguide, but this line's pretty long...

jamesob commented on src/test/validation_block_tests.cpp:
```diff
@@ -0,0 +1,158 @@
+
+    const std::shared_ptr<const CBlock> pblock = gen_invalid ? BadBlock(root) : GoodBlock(root);
+    blocks.push_back(pblock);
+    if (!gen_invalid)
```
Braces needed.

Empact replied:
clang-format would put those in I believe

jamesob reviewed (commented):
Concept ACK, just a few nits. Nice test!

Empact commented on src/test/validation_block_tests.cpp:
```diff
@@ -0,0 +1,158 @@
+    boost::thread_group threads;
+    for (int i = 0; i < 10; i++) {
+        threads.create_thread([&blocks]() {
+            bool blah;
```
Calling this `ignored` might be more straightforward

skeees commented:
Updated to address reviewer comments and fix a bug that this test seems to have uncovered

ryanofsky commented on src/validation.cpp:
```diff
@@ -2673,9 +2673,7 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
-            if (pindexMostWork == nullptr) {
-                pindexMostWork = FindMostWorkChain();
-            }
+            pindexMostWork = FindMostWorkChain();
```
Might be good to add a comment here saying why it's important to recompute pindexMostWork each loop iteration (how sync could get stuck otherwise).

sdaftuar replied:
This might cause substantial slow-down during IBD or reindex, because currently FindMostWorkChain walks from the candidate tip to the fork point on the current chain to ensure none of the blocks are invalid.  We may be able to optimize that, but we also might want to consider another solution.

ryanofsky commented on src/test/validation_block_tests.cpp:
```diff
@@ -0,0 +1,166 @@
+    pubKey << i++ << OP_TRUE;
+
+    auto ptemplate = BlockAssembler(Params()).CreateNewBlock(pubKey, false);
+    CBlock* pblock = new CBlock(ptemplate->block);
```
I'd think this could be written more simply as:

```
auto block = make_shared<CBlock>(ptemplate->block);
block->hashPrevBlock = ...
return block;
```

It looks like pblock is leaked currently, or is this not the case?

ryanofsky reviewed (commented):
utACK 6094ce73045fe0b4654ff94327c2059512af88fb. Fun tests!

sdaftuar commented:
Thanks for opening this PR.  I thought it would be helpful for other reviewers to clarify the (IMO significant) bug here, so that we can properly evaluate fixes.

If ActivateBestChain is invoked simultaneously in separate threads, then we can end up at a lower-work tip, and remain stuck until the next block comes in.

 * Suppose in thread 1, we have just been delivered a block that causes us to try to activate two blocks in ABC.  We connect 1 of them in ABCStep, and then release cs_main before connecting the second.
 * In thread 2, suppose we have been delivered a 3rd block (say via rpc) that builds on the first two.  It invokes ABC and gets to run after the first block has been connected in thread 1.  It connects one block, releases cs_main, and then connects one more, and finishes.
 * When thread 1 gets to run again, the most work chain has advanced, but (before this PR) we don't refresh pindexMostWork (except when we find an invalid block).  Consequently we would invoke ABCStep with a *less work tip* than our current tip(!).  This would cause us to disconnect our tip and return.

Some travis failures have been observed due to this bug, as seen for instance here: https://travis-ci.org/bitcoin/bitcoin/jobs/370848272.  The test that sometimes fails, `rpc_deprecated.py`, generates blocks on two nodes roughly simultaneously, so one of the nodes is generating blocks in an rpc thread while also processing blocks on the network thread, which I believe is enough to trigger this bug.

sdaftuar commented on src/test/validation_block_tests.cpp:
```diff
@@ -0,0 +1,166 @@
+    auto pblock = Block(prev_hash);
+
+    // a second coinbase will make this block invalid
+    pblock->vtx.push_back(pblock->vtx[0]);
```
This method of constructing a bad block is not ideal, because this kind of invalidity is detected well before ConnectBlock.  In order to get better code coverage, I think it would be better to generate different kinds of invalid blocks, to test failure at different points in the validation process.

(As it is,  I believe this method of making an invalid block is detectable in CheckBlock (merkle-tree fails validation due to duplicate transactions).  It would also fail in CheckBlock for having more than one coinbase, and it would fail in ContextualCheckBlock for having an invalid witness commitment.)

sdaftuar reviewed (changes requested):
Nice work finding this issue and good start on the new test.  I haven't finished my review yet, but here's my initial feedback.

sdaftuar commented:
@TheBlueMatt and I discussed that idea offline (of rewriting `FindMostWorkChain` to use `g_failed_blocks`, our internal data structure for caching invalid blocks, to speed it up) but our initial reaction was that the review overhead to ensure correctness might be very high.  I'm open to other ideas though...

skeees commented:
@sdaftuar - pulled in your commits.

Its still hard for me to convince myself that the set of conditions you've defined are the only cases in which `pindexMostWork` needs to be refreshed. Maybe I'm just not familiar enough with those areas - I'll leave it to the people with more experience in that area to review.

Just wanted to throw out one alternate way to do determine when to refresh most work tip that's (imo) a bit easier to convince myself behaves correctly. Instead of caching `pindexOldTip` and using that as well as a couple of other things to determine whether to refresh - how about giving setBlockIndexCandidates a sequence number that gets incremented every time it gets updated. Then all ABC has to do is record the last tip candidate set version it saw (instead of the old tip) and use that to determine whether a refresh may be necessary. You could do this in ABC or even internally in FindMostWorkChain.

MarcoFalke commented:
Copy of travis output:

```
Running tests: validation_block_tests from test/validation_block_tests.cpp
Running 1 test case...
Test cases order is shuffled using seed: 1374948318
Entering test module "Bitcoin Test Suite"
test/validation_block_tests.cpp(24): Entering test suite "validation_block_tests"
test/validation_block_tests.cpp(125): Entering test case "processnewblock_signals_ordering"
test/validation_block_tests.cpp(160): info:  has passed
test/validation_block_tests.cpp(160): info: ProcessNewBlock(Params(), block, true, &ignored)
test/validation_block_tests.cpp(160): info: check 
test/validation_block_tests.cpp(160): info: ProcessNewBlock(Params(), block, true, &ignored)
test/validation_block_tests.cpp(160): info: ProcessNewBlock(Params(), block, true, &ignored)
test/validation_block_tests.cpp(160): info: check 
test/validation_block_tests.cpp(160): info:  has passed
test/validation_block_tests.cpp(160): info: ProcessNewBlock(Params(), block, true, &ignored)
terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_S_create
make[3]: *** [test/validation_block_tests.cpp.test] Error 1

sdaftuar commented:
From some quick web searching, it seems that BOOST_CHECK may not be thread safe, so invoking it in each thread is likely the problem here.  Perhaps we should just `assert()` on the return value from `ProcessNewBlock()` (at line 160).

sdaftuar commented:
I do occasionally see this failure when I run validation_block_tests in a tight loop (after fixing the BOOST_CHECK issue above):
```
test/validation_block_tests.cpp(38): error in "processnewblock_signals_ordering": check m_expected_tip == block->hashPrevBlock failed [0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206 != 0000000000000000000000000000000000000000000000000000000000000000]
Segmentation fault (core dumped)
```

skeees commented:
Hmmm thats the genesis block which connects through a different path I believe

skeees commented:
@sdaftuar - thats the genesis block getting activated. Feels like it may be more of a test artefact (wasn't there a recent change that randomizes test case execution order) than a threading issue?  I think solution for that is to just call PNB(genesis) at the very beginning.

sdaftuar commented on src/test/validation_block_tests.cpp:
```diff
@@ -0,0 +1,179 @@
+    ProcessNewBlock(Params(), std::make_shared<CBlock>(Params().GenesisBlock()), true, &ignored);
+
+    // subscribe to events (this subscriber will validate event ordering)
+    TestSubscriber sub(chainActive.Tip()->GetBlockHash());
```
As discussed on IRC, there's a race condition where callbacks scheduled prior to a client subscribing to callbacks can be delivered to those clients, if the scheduler didn't finish draining its queue before the new subscriber is added.

That causes this test to rarely fail, because we initialize the TestSubscriber with the genesis block, even though it might be notified of the genesis block's connection (which has already happened at this point!) if the callbacks were delayed.  This can be reproduced by adding a `MilliSleep()` call to `SingleThreadedSchedulerClient::ProcessQueue()`.

The call to `ProcessNewBlock` at line 141 above doesn't fix this issue; instead we need to ensure that there are no outstanding callbacks at the time the TestSubscriber is created, for instance by calling `SyncWithValidationInterfaceQueue()`.  Also we should hold `cs_main` before accessing `chainActive`:

```
    SyncWithValidationInterfaceQueue();

    const CBlockIndex *initial_tip = nullptr;
    {
        LOCK(cs_main);
        initial_tip = chainActive.Tip();
    }
```

sipa reviewed (commented):
utACK code changes; I didn't review the tests.

Also, when introducing a test to demonstrate the problem, put it after the fix. Doing otherwise means introducing temporarily broken commits, which may break the ability to bisect.

sdaftuar commented on src/test/validation_block_tests.cpp:
```diff
@@ -0,0 +1,171 @@
+
+#include <iomanip>
+#include <memory>
+#include <sstream>
```
nit: doesn't seem like these includes are necessary at all, am I missing some dependency?

sdaftuar reviewed (approved):
ACK c496b7b386a66fb4fc73d3bacd3db2a5c4bf7872, though I left some nits for you to consider.

skeees commented:
@sipa - commits now correctly ordered and `m_cs_cs` renamed to `m_cs_chainstate`

sdaftuar commented:
re-ACK ed1828c202c9f4376492c173e361814320a01ffa

skeees committed:
Fix concurrency-related bugs in ActivateBestChain

If multiple threads are invoking ActivateBestChain, it was possible to have
them working towards different tips, and we could arrive at a less work tip
than we should.  Fix this by introducing a ChainState lock which must
be held for the entire duration of ActivateBestChain to enforce
exclusion in ABC.

jamesob commented on src/validation.cpp:
```diff
@@ -2666,38 +2679,47 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
+                    GetMainSignals().BlockConnected(trace.pblock, trace.pindex, trace.conflictedTxs);
+                }
+            } while (!chainActive.Tip() || (starting_tip && CBlockIndexWorkComparator()(chainActive.Tip(), starting_tip)));
+            if(!blocks_connected) return true;
```
nit: `if (...`

jamesob commented on src/validation.cpp:
```diff
@@ -144,6 +144,12 @@ class CChainState {
 
+    /**
+     * the ChainState CriticalSection
+     * A lock that must be held when modifying this ChainState - held in ActivateBestChain()
```
> must be held when modifying this ChainState

Seems like this statement is too broad; many aspects of this object are modified without this lock held (e.g. `g_failed_blocks`, `setBlockIndexCandidates`).

sdaftuar replied:
I agree -- this is more of a lock on the logic in ABC than any particular data structure.  The comment in ABC itself is pretty good though.

jamesob commented on src/validation.cpp:
```diff
@@ -2653,6 +2660,12 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
     // sanely for performance or correctness!
     AssertLockNotHeld(cs_main);
 
+    // ABC maintains a fair degree of expensive to calculate internal state
```
I had a bit of trouble parsing this comment initially - might be easier to read if you wrote "expensive-to-calculate" instead?

jamesob reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/pull/13023/commits/8febf4850b11406883d457d3c63e713acfacf682

The tests you've written are great (we need more like these) and the changeset fixes a definite bug, so I'm :+1: on this. I left a few nits but feel free to ignore or fix later.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2666,38 +2666,47 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
-                assert(trace.pblock && trace.pindex);
-                GetMainSignals().BlockConnected(trace.pblock, trace.pindex, trace.conflictedTxs);
-            }
+            const CBlockIndex * pindexFork = chainActive.FindFork(pindexOldTip);
```
Not sure if this is my fault or not, but this definitely needs to be finding the fork since starting_tip for the notifications, not just since the last loop iteration. Also, we should probably move the UpdatedBlockTip callback generation inside the pindexFork != pindexNewTip check as otherwise you'll get rather nonsense callbacks (and none of the callers seem to rely on any specific behavior). Either way behavior needs documentation in validationinterface.h (or I can do it in #12979 once this gets merged).

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2653,6 +2660,12 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
 
+    // ABC maintains a fair degree of expensive to calculate internal state
+    // because this function periodically releases cs_main so that it does not lock up other threads for too long
+    // during large re-orgs - and to allow for e.g. the callback queue to drain
```
nit: comment is slightly wrong now, we dont release cs_main during large reorgs, only large connects.

TheBlueMatt commented on src/test/scheduler_tests.cpp:
```diff
@@ -112,4 +112,46 @@ BOOST_AUTO_TEST_CASE(manythreads)
+    }
+
+    // these are not atomic, if SinglethreadedSchedulerClient prevents
+    // parallel execution at the queue level no synchronization should be required here
```
This isnt sufficient - just because two things are ordered does not mean they have a consistent view into memory unless there was something that triggered a flush. In practice I believe the lock inside of the SingleThreadedSchedulerClient provides this, but that's certainly not guaranteed in the future. A release-acquire should be sufficient, however.

skeees replied:
don't quite understand what you mean here - isn't the whole purpose of SingleThreadedSchedulerClient to ensure internally consistent ordering/views of callbacks on a single client? if a memory barrier is somehow necessary to achieve this (like you I think the lock is sufficient for now) then that should be done in the SingleThreadedSchedulerClient and not be the caller's responsibility. This test was intended to demonstrate that the api guarantees this behavior to users.

TheBlueMatt replied:
The C++ memory model does not provide any ordering guarantees across variables at different memory locations unless you have a) some operation dependency tree which creates such an order, or b) a seq_cst operation somewhere which implicitly creates such an order (which a mutex lock/unlock essentially implicitly creates). So if, eg, the SingleThreadedSchedulerClient used only release/acquire operations to load the next callback to call, and both callback clients update the variable, there is no happens-before relationship created on the variable and you may not have consistent results. Having the seq_cst implied by the internal lock provides what you want implicitly, but its not a guarantee. A release/acquire operation should be sufficient.

skeees replied:
oh hmm - but each client only interacts with its own "counter" there's no sharing across clients in this tests. I'm simply checking that for each client, its own callbacks are processed sequentially - which i think is something that the singlethreadedschedulerclient api is supposed to guarantee (otherwise might as well just use the raw scheduler) - and no current test existed

TheBlueMatt replied:
Ugh, my memory of the C++ memory model was slightly faulty, I've updated the comment above, though its much less a concern now (however should sitll be fixed). You could use relaxed operations to try to optimistically fetch the next callback to execute, succeed, and then have one thread run a callback, the other thread run a second callback but for the same client, and create no dependancy. On many non-x86 platforms the CPU is free to reorder (apparently) unrelated stores/loads.

TheBlueMatt reviewed (commented):
Didnt review the last new test (awesome, btw).

TheBlueMatt commented:
utACK non-test changes through a3ae8e68739023e5dba9e5cb190e707ed4603316, a3070a861bb3a76303b7859c4dd96ba886029111 would be fine with some release/acquire atomics, and 51539506577e5439c58b2e36c07d84f33bb4d4c5 looks awesome, though I haven't reviewed it fully.

skeees commented:
I will follow up offline in case I'm totally missing something - but I feel pretty strongly that sequential consistency should be guaranteed by the `SingleThreadedSchedulerClient` - whether the callbacks are actually run by one or multiple threads internally, it should feel single threaded (it's even named that way) and the caller shouldn't have to worry about it. Otherwise you open up a ton of surface area for more bugs like the ones being fixed here to creep in.
Of course if there's something I'm missing in the current implementation of the `SingleThreadedSchedulerClient` that doesn't guarantee this ordering, I'm happy to fix it here, but it appears that the lock guarantees it now and I'd strongly advocate that any future implementations without the lock still maintain that guarantee.

TheBlueMatt commented:
@skeees hmm, maybe we're talking past each other, but my point was that you can fully guarantee both ordering and only one callback executing at once *without* implying seq_cst/rel_acq in the C++ memory model. I dont think the memory consistency of a client should be implied by an API.

skeees commented:
Ok - I finally understand what you're staying. I still disagree. I think it makes the API pretty unusable if you drop that guarantee - and the theoretical (because its single threaded right now) performance tradeoff is worth the usability - and certainly it should not be called SingleThreaded anything since it drops the single threaded memory model. But in the interest of not delaying 0.16.1 I've removed that commit and will follow up in a separate PR.

TheBlueMatt commented:
utACK 7b4be5089938b188ecca86742aa6aa4d2ec0c40e, though I didn't review the last test in detail (thanks for making that!).

laanwj commented:
Compiling this locally gives me:
```
/.../bitcoin/src/test/validation_block_tests.cpp:73:13: error: use of undeclared identifier 'CheckProofOfWork'
    while (!CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())) {
            ^
1 error generated.
```

Adding `#include <pow.h>` fixes it.

skeees commented:
thanks @laanwj - updated

sdaftuar reviewed (commented):
Post-merge ACK, with some nits.

fanquake commented:
Backported in #13317
